File: inev_sdk/utils/state_inference.py
"""
State inference from HTTP context.

Provides zero-config state inference from HTTP method and status code
combinations. Uses standard REST conventions to determine the resulting
state of an operation.

State Mapping:
    POST + 201 → "created"
    POST + 200 → "processed" (non-creation POST like action endpoints)
    PUT + 200/204 → "updated"
    PATCH + 200/204 → "updated"
    DELETE + 200/204 → "deleted"
    GET + 200 → None (reads don't change state)
    * + 4xx/5xx → None (errors handled via outcome field)
"""

from __future__ import annotations


def infer_state_from_http(
    method: str,
    status_code: int,
    action: str | None = None,
) -> str | None:
    """Infer to_state from HTTP method and status code.

    Uses REST conventions to determine the resulting state of an operation.
    Only infers state for successful write operations.

    Args:
        method: HTTP method (GET, POST, PUT, PATCH, DELETE)
        status_code: HTTP response status code
        action: Optional action name for additional context

    Returns:
        Inferred state name, or None if state cannot be determined.

    Examples:
        >>> infer_state_from_http("POST", 201)
        "created"

        >>> infer_state_from_http("PUT", 200)
        "updated"

        >>> infer_state_from_http("DELETE", 204)
        "deleted"

        >>> infer_state_from_http("GET", 200)
        None  # Reads don't change state

        >>> infer_state_from_http("POST", 500)
        None  # Errors handled via outcome
    """
    method = method.upper()

    # Don't infer state for error responses (4xx, 5xx)
    if status_code >= 400:
        return None

    # Don't infer state for redirects (3xx)
    if status_code >= 300:
        return None

    # GET requests don't change state
    if method == "GET":
        return None

    # HEAD and OPTIONS don't change state
    if method in ("HEAD", "OPTIONS"):
        return None

    # POST - typically creates, but could be action endpoint
    if method == "POST":
        if status_code == 201:
            return "created"
        if status_code == 200:
            # POST with 200 could be many things - check action for hints
            if action:
                return _infer_state_from_action(action)
            return "processed"
        if status_code == 202:
            return "pending"  # Accepted for async processing
        return None

    # PUT - replaces/updates resource
    if method == "PUT":
        if status_code in (200, 204):
            return "updated"
        if status_code == 201:
            return "created"  # PUT can create if resource doesn't exist
        return None

    # PATCH - partial update
    if method == "PATCH":
        if status_code in (200, 204):
            # Check if action gives more specific state
            if action:
                inferred = _infer_state_from_action(action)
                if inferred:
                    return inferred
            return "updated"
        return None

    # DELETE - removes resource
    if method == "DELETE":
        if status_code in (200, 204):
            return "deleted"
        if status_code == 202:
            return "deleting"  # Async deletion
        return None

    return None


def _infer_state_from_action(action: str) -> str | None:
    """Infer state from action name patterns.

    Looks for common verb prefixes in action names to determine state.

    Args:
        action: Action name (e.g., "post_order_confirm", "patch_user_activate")

    Returns:
        Inferred state or None
    """
    if not action:
        return None

    action_lower = action.lower()

    # Check for state-changing verbs in the action name
    state_verbs = {
        "confirm": "confirmed",
        "approve": "approved",
        "reject": "rejected",
        "cancel": "cancelled",
        "archive": "archived",
        "activate": "activated",
        "deactivate": "deactivated",
        "suspend": "suspended",
        "resume": "resumed",
        "complete": "completed",
        "submit": "submitted",
        "publish": "published",
        "unpublish": "unpublished",
        "verify": "verified",
        "enable": "enabled",
        "disable": "disabled",
        "start": "started",
        "stop": "stopped",
        "pause": "paused",
        "ship": "shipped",
        "deliver": "delivered",
        "refund": "refunded",
        "expire": "expired",
        "renew": "renewed",
        "upgrade": "upgraded",
        "downgrade": "downgraded",
    }

    # Sort by verb length (longest first) to handle overlapping verbs
    # e.g., "deactivate" must be checked before "activate"
    for verb, state in sorted(state_verbs.items(), key=lambda x: len(x[0]), reverse=True):
        if verb in action_lower:
            return state

    return None


def infer_from_state_by_method(method: str) -> str | None:
    """Infer from_state based on HTTP method semantics.

    For certain methods, we can infer what state the resource was likely in
    before the operation.

    Args:
        method: HTTP method

    Returns:
        Inferred from_state or None

    Note:
        This is less reliable than to_state inference and should be used
        with caution. State tracking is more accurate for from_state.
    """
    method = method.upper()

    # DELETE implies resource existed (was in some active state)
    if method == "DELETE":
        return "active"  # Generic "was active before deletion"

    # Can't reliably infer from_state for other methods
    return None

File: go-sdk/utils/state_inference.go
package utils

import (
	"strings"
)

// StateTransition represents a from/to state pair.
type StateTransition struct {
	FromState string
	ToState   string
}

// InferState infers the state transition based on HTTP method and status code.
// Returns the to_state (from_state is typically not inferable without prior context).
func InferState(method string, statusCode int) string {
	method = strings.ToUpper(method)

	// Error status codes don't result in state changes
	if statusCode >= 400 {
		return ""
	}

	switch method {
	case "POST":
		switch statusCode {
		case 201:
			return "created"
		case 200, 202:
			return "processed"
		default:
			return ""
		}
	case "PUT":
		if statusCode == 200 || statusCode == 204 {
			return "updated"
		}
	case "PATCH":
		if statusCode == 200 || statusCode == 204 {
			return "updated"
		}
	case "DELETE":
		if statusCode == 200 || statusCode == 204 || statusCode == 202 {
			return "deleted"
		}
	case "GET":
		// GET requests don't change state
		return ""
	}

	return ""
}

// InferStateFromAction infers state from the action name.
// This handles cases where the action verb indicates the target state.
func InferStateFromAction(action string) string {
	action = strings.ToLower(action)

	// Extract the verb from action names like "confirm_order", "post_order_confirm"
	parts := strings.Split(action, "_")

	// Check each part for a state-indicating verb
	stateVerbs := map[string]string{
		// Lifecycle states
		"create":    "created",
		"created":   "created",
		"update":    "updated",
		"updated":   "updated",
		"delete":    "deleted",
		"deleted":   "deleted",
		"remove":    "removed",
		"removed":   "removed",
		"archive":   "archived",
		"archived":  "archived",
		"unarchive": "unarchived",
		"restore":   "restored",

		// Approval workflow
		"confirm":   "confirmed",
		"confirmed": "confirmed",
		"approve":   "approved",
		"approved":  "approved",
		"reject":    "rejected",
		"rejected":  "rejected",
		"cancel":    "canceled",
		"canceled":  "canceled",

		// Activation states
		"activate":   "activated",
		"activated":  "activated",
		"deactivate": "deactivated",
		"suspend":    "suspended",
		"suspended":  "suspended",
		"resume":     "resumed",
		"resumed":    "resumed",
		"enable":     "enabled",
		"enabled":    "enabled",
		"disable":    "disabled",
		"disabled":   "disabled",

		// Progress states
		"start":     "started",
		"started":   "started",
		"stop":      "stopped",
		"stopped":   "stopped",
		"pause":     "paused",
		"paused":    "paused",
		"complete":  "completed",
		"completed": "completed",
		"finish":    "finished",
		"finished":  "finished",

		// Submission states
		"submit":    "submitted",
		"submitted": "submitted",
		"publish":   "published",
		"published": "published",
		"unpublish": "unpublished",
		"draft":     "draft",

		// Verification states
		"verify":     "verified",
		"verified":   "verified",
		"validate":   "validated",
		"invalidate": "invalidated",

		// E-commerce states
		"ship":      "shipped",
		"shipped":   "shipped",
		"deliver":   "delivered",
		"delivered": "delivered",
		"refund":    "refunded",
		"refunded":  "refunded",
		"pay":       "paid",
		"paid":      "paid",

		// Expiration states
		"expire":   "expired",
		"expired":  "expired",
		"renew":    "renewed",
		"renewed":  "renewed",
		"extend":   "extended",
		"extended": "extended",

		// Upgrade states
		"upgrade":    "upgraded",
		"upgraded":   "upgraded",
		"downgrade":  "downgraded",
		"downgraded": "downgraded",

		// Assignment states
		"assign":   "assigned",
		"assigned": "assigned",
		"unassign": "unassigned",
		"transfer": "transferred",

		// Lock states
		"lock":     "locked",
		"locked":   "locked",
		"unlock":   "unlocked",
		"unlocked": "unlocked",
	}

	for _, part := range parts {
		if state, ok := stateVerbs[part]; ok {
			return state
		}
	}

	return ""
}

// InferFullTransition attempts to infer both from and to states.
// This is most useful when the action contains the state verb.
func InferFullTransition(method string, statusCode int, action string) StateTransition {
	trans := StateTransition{}

	// Try to infer to_state from action first
	if action != "" {
		trans.ToState = InferStateFromAction(action)
	}

	// Fall back to HTTP method + status code
	if trans.ToState == "" {
		trans.ToState = InferState(method, statusCode)
	}

	return trans
}

// IsSuccessStatus returns true if the status code indicates success.
func IsSuccessStatus(statusCode int) bool {
	return statusCode >= 200 && statusCode < 300
}

// IsClientError returns true if the status code indicates a client error.
func IsClientError(statusCode int) bool {
	return statusCode >= 400 && statusCode < 500
}

// IsServerError returns true if the status code indicates a server error.
func IsServerError(statusCode int) bool {
	return statusCode >= 500
}
